lib/nearley.js

```diff
@@ -338,6 +337,73 @@
         return this;
     };
 
+
+    Parser.prototype.reportError = function(token) {
+        var lines = [];
+        const tokenDisplay = (token.type ? token.type + " token: " : "") + JSON.stringify(token.value !== undefined ? token.value : token);
+        lines.push(this.lexer.formatError(token, "Syntax error"));
+        lines.push(`Unexpected ${tokenDisplay}. Instead, I was expecting to see one of the following:\n`);
+        const lastColumnIndex = this.table.length - 2;
+        const lastColumn = this.table[lastColumnIndex];
+        // Display each state that is expecting a terminal symbol next.
+        for (let i = 0; i < lastColumn.states.length; i++) {
+            const state = lastColumn.states[i];
+            const nextSymbol = state.rule.symbols[state.dot];
+            if (nextSymbol && typeof nextSymbol !== "string") {
+                const symbolDisplay = this.getSymbolDisplay(nextSymbol);
+                lines.push(`    A ${symbolDisplay} based on:`);
+                // Display the "state stack" - which shows you how this state
+                // came to be, step by step.
+                const stateStack = this.buildStateStack(state);
+                for (let j = 0; j < stateStack.length; j++) {
+                    const state = stateStack[j];
+                    lines.push(`        ${state.rule.toString(state.dot)}`);
+                }
+            }
+        }
+        lines.push("");
+        return lines.join("\n");
+    };
+
+    Parser.prototype.getSymbolDisplay = function(symbol) {
+        const type = typeof symbol;
+        if (type === "string") {
+            return symbol;
+        } else if (type === "object" && symbol.literal) {
+            return JSON.stringify(symbol.literal);
+        } else if (type === "object" && symbol instanceof RegExp) {
+            return `character matching ${symbol}`;
+        } else {
+            throw new Error(`Unknown symbol type: ${symbol}`);
+        }
+    };
+
+    /*
+    Builds the "state stack" - which you can think of as the call stack of the
+    recursive-descent parser which the Nearley parse algorithm simulates.
+    The state stack is represented as an array of state objects. This function
+    needs to be given a starting state, and it returns an array of state objects. 
+    The first item of the array will bet the starting state, with each successive
+    item in the array going further back into history.
+    */
+    Parser.prototype.buildStateStack = function(state) {
+        if (!state) {
+            return [];
+        }
+        // console.log(`${state}, wantedBy: ${state.wantedBy.join(", ")}`)
+        if (state.dot === 0) {
+            return [
+                state,
+                ...this.buildStateStack(state.wantedBy[0])
```

Comment:
Can this be de-sugared out of the `...` syntax? We've tried to keep nearley's core in "old-style" JS throughout and this would be a weird deviation.

Reply:
Done.